docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,187 @@
+
+## Motivation
+
+Deployment scenarios with various levels of federated setups are not unusual. Concrete deployments i.e. leverage Thanos Ruler to spawn multiple query sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and Thanos Ruler instances.
+
+Pitfalls of current solutions:
+* It’s tedious to manually visit each leaf and we are lazy.
+* Visiting leaf instances (be it Thanos Ruler or dedicated Prometheus endpoints) have to be implemented manually.
+
+## Goals
+
+* Present other Prometheus/Thanos Resources in a global view.
+    * This also means an up-to-date view (e.g statuses)
+* Simple to run if you already run Thanos.
+    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
+    * Potentially use the same connections, discovered targets by Querier.
+
+## Verification
+
+* Unit tests which would fire up a dummy Rules API and check different scenarios.
+* Ad-hoc testing.
+
+## Proposal
+
+### Rules API
+
+We propose to add following proto service and propagate all using existing components that has rules, so:
+* Sidecar
+* Querier (for federation)
+* Ruler
+
+The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alerting rules.
+
+```proto
+service Rules {
+  /// Rules has info for all rules.
+  rpc Rules(RulesRequest) returns (stream RulesResponse);
+}
+```
+
+Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly, Thanos Ruler constructs the response based in its local state.
+
+Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:
+
+* Deduplication happens first the rule group level. The identifier is the group name.
+* Then, per group name deduplication happens on the rule level, where:
+
+1. the rule type (recording rule vs. alerting rule)
+2. the rule name
+3. the rule label names
+
+are being used as the deduplication identifier. Disjunct entries are simply merged by adding them to the result set.
+
+Thanos query presents the result of the fan-out on a `/api/v1/rules` endpoint which is compatible with the Prometheus' rules API endpoint.
+Additionally Thanos Querier gains a new `--rule.replica-label` command line argument which falls back to `--query.replica-label` if unset.
```

Comment:
Could we also describe how are we going to set the replica label in sidecar and thanos ruler?

Right now when we do Prometheus metric dedup, we set it in `external_labels` of prometheus config:
```
  external_labels:
    replica: replica-$(HOSTNAME)
```

Reply:
This is too complex.. Why we cannot have just `query` replica and only that?

Reply:
so we can reuse external labels for sidecar and thanos rule has `--label` already?

Reply:
Yup, great idea. Let's add this to the proposal.

Reply:
ok, sounds good, will update the design doc :+1:

Reply:
Added a small section about Thanos Ruler `--label` and `external_labels` that will be reused. How about Thanos Querier though? As far as I see it does not have a `--label` command line argument, I guess we need one here too?